[Sky] 미션5 : 보드에 위치 부여 및 점수 계산

aksmf1442 opened the pull request:
안녕하세요! 미션 5의 요구사항을 따라가다보니 미션4에 설계했던 코드랑 많이 달라져서 이상한 부분이 많더라도 리뷰 잘 부탁드립니다! 

## 미션 5 주요 변경사항

* 미션5의 다섯 번째 요구사항까지는 List 자료구조를 통해서 initialize한 기물들을 관리하고 있었는데 여섯 번째 요구사항에서 주어진 위치의 기물을 조회할 수 있도록 메소드를 구현하라고 나와있어서 이건 Map 자료구조를 사용해서 Map<Position, Piece>와 같이 관리하는게 더 편하겠다 싶어서 주어진 위치의 기물을 조회하는 기능은 map을 따로 만들어서 조회할 수 있도록 만들었습니다.

* 위에서 말한 것처럼 map을 만들긴 했었지만 위치를 조회하는 용도로만 사용하고 다른 곳에서는 사용하지 않고 있었는데 조원들과의 그룹리뷰 시간에 토론을 하면서 Map 자료구조를 사용할 거면 list 자료구조를 굳이 사용할 필요가 없다는 것을 깨닫고, list를 전부 없애고 map을 통해서 모든 것을 할 수 있도록 리팩토링 했습니다.

* 체스 프로그램 점수 계산하는 요구사항에서는 같은 색의 pawn이 같은 file에 두 개 이상 있으면 그 file에 있는 pawn의 점수가 0.5점이 되는 부분이 까다로웠는데, 이 부분은 Piece 클래스의 getPoint()에 pawn이 있는 file의 리스트를 넘겨줘서 그 file리스트에 pawn이 두 개 이상있으면 0.5점을 반환하고 아니면 1점을 반환할 수 있도록 했습니다.

aksmf1442 committed:
Add NOCOLOR constatnt to Color enum, Add No_PIECE to PieceType enum

비어있는 체스판도 관리하기 위해서 Color enum에 NOCOLOR 상수추가, PieceType enum에 No_PIECE를 추가했다.
blank 객체를 생성하기 위해서 PieceMaker에 createBlank()를 추가했는데, 이 경우에는 색깔이 NOCOLOR라서 static 메서드로 추가했다.

aksmf1442 committed:
Add Rank class for board list, Refactor Board class

기물들의 위치이동을 원활하게 하기 위해서 Rank 클래스를 만든 후 Board 클래스에 board 리스트를 만들고 rank를 관리 할 수 있도록 했다.

aksmf1442 committed:
Add pieceCount method to Board, Rank

기물과 색에 해당하는 기물의 개수를 반환하기 위해서 Board와 Rank 클래스에 pieceCount()를 추가했다.

aksmf1442 committed:
Add findPiece method to Board

position 문자열을 파라미터로 받고 board의 position에 위치한 piece를 찾기 위해서 findPiece()를 추가했다.
또한 BoardTest에서 findPiece()로 찾은 객체와 비교할 객체의 equals 비교를 위해 Piece 클래스에 equals(), hashcode()를 오버라이딩 했다.

aksmf1442 committed:
Add Position, Refactor pieceFind method of Board

위치를 편하게 관리하기 위해서 Position클래스를 추가하고 Board클래스에 map<Position, Piece>와 기물 위치정보를 관리하게 하도록 변경했다.

aksmf1442 committed:
Refactor Position, Board, Add RankMaker

Position 클래스의 x, y변수에 값이 잘못 넣어져서 수정했다.
기존의 Board클래스에서 initialize 메서드 로직들이 중복되는 부분들이 많아져서 RankMaker라는 클래스를 새로 만들고,
이 곳에서 Rank를 만들 수 있게 해줌으로써 Board의 initialize 메서드 로직들을 대폭 줄여줬다.

aksmf1442 committed:
Add move method to Board

임의의 기물을 체스판 위에서 이동(추가)할 수 있도록 Board클래스에 move()를 추가했다.

aksmf1442 committed:
Remove : Rank, RankMaker class, Feat : calculatePoint to Board class

순서대로 미션을 따라오다가 map을 사용할꺼면 굳이 list로 rank를 관리할 필요성이 없겠다고 느껴서 Rank, RankMaker 클래스들을 제거하고,
Board 클래스의 list들을 전부 제거한 후 그로 인해서 일어나는 사이드 이펙트를 전부 리팩토링했다.
Board에 색깔별 기물들의 점수를 더해주는기능을 추가했다.

aksmf1442 committed:
Refactor : rename variable to Board class, Feat : getPoint to Piece

Board 클래스 변수들의 이름을 조금 더 알기 쉽게 변경했다.
기물의 점수를 받아올 수 있도록 Piece 클래스에 getPoint()를 추가했다.

aksmf1442 committed:
Feat : sortByScore function

색깔별 기물의 점수가 높은 순으로 정렬하는 기능을 만들기 위해서 Board 클래스에 sortByScore 메서드를 추가했다.

aksmf1442 committed:
Refactor : variable name

전체적으로 변수 네이밍을 다른사람이 봐도 조금이나마 더 잘 알아볼 수 있도록 변경했다.

wheejuni commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -3,107 +3,159 @@
+    private final int END_RANK = 1;
+    private final char START_FILE = 'a';
+    private final char END_FILE = 'h';
+    private Map<Position, Piece> piecePositions = new LinkedHashMap<>();
```
`LinkedHashMap` 을 선택하신 이유가 궁금하네요.

aksmf1442 replied:
순서대로 화면에 출력하고 싶어서 LinkedHashMap을 사용했습니다! 그런데 생각해보니까 출력할 때도 LinkedHashMap을 사용하지 않고 순서대로 출력할 수 있는 방법이 있네요. 다시 HashMap으로 변경해서 사용해 봐야겠네요.

wheejuni commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -3,107 +3,159 @@
-            addWhitePawn(whitePieceMaker.createPawn());
+    public String showBoard() {
+        StringBuilder result = new StringBuilder();
+        final int FILE_SIZE = 8;
```
이 변수는 끌올해서 클래스 상단부에 선언돼야 할 것 같아요. 
상수 성격의 변수가 지역변수로 선언되어 있는 것은 아무리 봐도 조금은 이상한데요. 어떻게 생각하시나요?

aksmf1442 replied:
제가 아직 부족한 점이 많아 모르는것도 많아서 길게 질문 드립니다! 
1. 상수 성격의 변수가 메소드 한 곳에서만 쓰여도 무조건 클래스 상단부에 선언하는게 더 좋은 방법인가요?  
2. 그리고 사실 이 FILE_SIZE와 같은 값인 START_RANK라는 변수가 클래스 상단부에 있는데 제가 전달하고 싶은 의미와 변수명이 품은 의미가 달라서 새로 변수를 초기화 했는데, 이 경우엔 그냥 값은 같으니까 START_RANK라는 변수를 사용하는게 좋을까요 아니면 값이 같더라도 전달하고 싶은 의미의 변수를 새로 만드는게 좋을까요?
답변해주시면 감사하겠습니다!👍

wheejuni replied:
이 메소드 한 곳에서만 사용된다면 상수로 보지 않아도 되고, 이 경우에 네이밍 컨벤션은 일반 변수와 같이 카멜 케이스(`fileSize`) 로 해 주세요. 

값이 같아도 사용처와 의미가 다르다면 다른 변수로 선언해주세요.

aksmf1442 replied:
답변 감사합니다!

wheejuni commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -3,107 +3,159 @@
+    public int pieceCount(Piece piece) {
+        int pieceCount = 0;
+
+        for (Piece p : piecePositions.values()) {
+            if (p.equals(piece)) {
+                pieceCount += 1;
+            }
```
`stream` 으로 개선할 수 있을 것 같은데 어떨까요? 

```suggestion
        placePositions.values().stream().filter(p -> p.equals(piece)).count()

aksmf1442 replied:
훨씬 직관적인 것 같아서 좋은 것 같아요. 변경 해볼게요!

wheejuni commented on src/main/java/net/sky/chess/Board.java:
```diff
@@ -3,107 +3,159 @@
+            Piece piece = piecePositions.get(position);
+            char file = position.getFile();
+            List<Piece> fileList = getFile(file, color);
+            if (color == piece.getColor()) {
```
`.get()` 해서 꺼낸 후 비교하는 방법보다는 `Piece` 에게 색깔을 직접 물어보는 방법이 훨씬 안전하고 깔끔할 것 같아요.
아래와 같은 메소드가 `Piece` 에 구현되어 있으면 될 텐데요, 어떤가요?

```java
public boolean isMatchingColor(Color color);

aksmf1442 replied:
변경했습니다!

wheejuni commented on src/main/java/net/sky/pieces/Piece.java:
```diff
@@ -32,9 +36,53 @@ public boolean isBlack(Piece piece) {
+        return pieceType.getPoint();
+    }
+
+    public double getPoint(List<Piece> fileList) {
```
이 메소드는 여기에 위치하면 안될 것 같다는 생각이 듭니다. 
`Piece` 가 자신 여러개를 갖고 있는 `fileList` 를 파라메터로 받아서 직접 점수까지 계산하는 로직은 좀 어색합니다. 
`Piece` 는 자기 자신만의 점수를 관리하도록 하고 계산은 다른 곳에서 이뤄지는게 더 좋아 보이네요.

aksmf1442 replied:
생각해보니까 자기 자신을 여러 개를 파라미터로 받는 부분은 어색한 것 같네요. 변경했습니다!

wheejuni commented on src/main/java/net/sky/pieces/Piece.java:
```diff
@@ -1,5 +1,9 @@
+
+import java.util.List;
+import java.util.Objects;
+
```
갑자기 생각난 건데, `Piece` 가 `Position` 도 직접 들고 있는 구현은 어떨까요?
만약 그렇다면 `Board` 가 굳이 `LinkedHashMap` 의 형태로 관리할 필요도 없을 것 같은데요?

aksmf1442 replied:
그 구현도 해볼까 생각했었는데 위치를 찾고 변경하는게 Map이 더 편할 것 같아서 Map을 사용해봤습니다!

wheejuni replied:
`Map` 중에서도 `LinkedHashMap` 을 선택한 이유는요?

wheejuni commented on src/main/java/net/sky/pieces/Piece.java:
```diff
@@ -32,9 +36,53 @@ public boolean isBlack(Piece piece) {
+            }
+        }
+
+        return count > 1 ? defaultPoint / 2 : defaultPoint;
```
삼항연산자의 활용 자체는 나쁘지 않은데 개인적인 생각으론 삼항연산자 내부에 연산이 들어가면 그때부턴 가독성이 좀 떨어집니다. 
차라리 아래와 같이 풀어쓰는 것이 가독성 측면에서 더 나을수도 있겠습니다.

```suggestion
        if (count > 1) {
            return defaultPoint / 2;
        }
        return defaultPoint;

aksmf1442 replied:
변경했습니다!

wheejuni reviewed (changes requested):
5단계까지 진행하시느라 수고 많으셨습니다. 💯 
몇 가지 개선 의견 남겨드렸는데 참고해주세요. 감사합니다~

aksmf1442 committed:
Refactor : Board Class

보드를 출력할 때 순서대로 출력하기 위해서 LinkedHashpMap을 사용하고 있었는데 HashMap을 쓰더라도 순서대로 출력할 수 있도록
showBoard()를 변경해서 HashMap으로 관리하도록 했습니다.
기존의 Piece 클래스에서 getPoint() 메소드의 파라미터로 List<Piece>를 받는 부분이 어색해서 이 부분을 Board클래스로 빼고 그에 맞게
Board클래스의 메소드들을 추가하고 변경했습니다.

aksmf1442 commented:
사실 체스판 자체가 8x8이라 list나 배열로 관리하는게 맞는 것 같아서 처음에는 list로 관리를 하고 있었는데요..🥲  위치를 찾고 변경하는데 map을 사용하려니까 기존의 list와 새로 사용하려는 map 둘 다 쓰면 자원낭비?라는 느낌이 들었습니다. 그래서 체스의 핵심은 기물들의 위치이동이라고 생각을 해서 편하게 위치이동을 하기 위해서 list를 지우고 map만 남겨뒀는데요. 이 방법은 별로 좋지 않은 방법일까요?

wheejuni commented:
@aksmf1442 단순히 `List` 를 `Map` 으로 변경한 일에 대해 좋은 방법, 좋지 않은 방법임을 판단하기는 어렵네요.
여기에 의존되어있는 다른 로직들이 어떻게 작성되어있는지가 훨씬 더 중요한데요, 대체로 무난하게 구현된 듯 합니다.
